Add destination and temporary directory options

masupilamie opened the pull request:
As discussed via e-mail I've added two optional arguments to the script that allow usage of different directories for converted and temporary files. This change touches a lot of the code, so probably best to review. Tested with DV71 and HDR10+ files. This would close issue #21 and allows for conversion speedup by using different filesystems (hardware). It also enables usage when the source directory is read-only or should be left unchanged.

gacopl commented:
on a quick look dest dir is not applied in mp4box, dvconverting cannot be in tmp otherwise the whole thing stops working in multiple files dir. probably more, also default tempdir would be ok to have ./tmp

gacopl commented:
all it takes to have 2 triggers to run this script on same dir, imagine tv show season if dvconverting will be not in place things will break, because of all the tmp files going on, there cannot be 2 conversions running on same source dir

masupilamie commented:
I think I understand what you mean. In that case the script without my changes will already run into issues when it is executed multiple times in the same directory with multiple mkv files in it, since the dvconverting file uses the input variable. I suggest to make the .dvconverting file static so it locks the whole directory during script run (ie just .dvconverting without the filename as prefix).

I've made the dvconverting file static and added the option to move the lock file to temp, see last commit.

gacopl commented on dvmkv2mp4:
```diff
@@ -479,10 +526,17 @@ EOF
+
+rm "${LOCKFILE}"
+
+if [ "$TEMP_DIR_CREATED" = 'yes' ]; then
```
this should be moved to debug section, in case of debug we want to keep temp files to be able to debug conversion

gacopl commented on dvmkv2mp4:
```diff
@@ -405,9 +452,9 @@ EOF
   ### GRAB CHAPTERS
   if [[ "$SRC_FORMAT" == "mp4" ]]; then
-    MP4Box -dump-chap-ogg "${input}" && mv "$inputbase".txt chapters.list
+    MP4Box -dump-chap-ogg "${input}" -std > "${TEMP_DIR}chapters.list"
```
what this std does ?

gacopl replied:
nevermind :D

gacopl commented on dvmkv2mp4:
```diff
@@ -15,6 +15,12 @@ ADDSUBS="no"
+## Directory to store created files in, disabled (empty) by default. When specified and TEMP_DIR is not, this will also be used for temporary files.
+DEST_DIR=""
+## Directory to store temporary files in. Defaults to current directory. Script will create a 'tmp' subfolder in this directory for its temporary files.
+TEMP_DIR=""
```
to keep similar behaviour to previois this should be ./tmp by default

gacopl replied:
nvm i see it now

gacopl commented on dvmkv2mp4:
```diff
@@ -111,11 +121,8 @@ function cleanup {
     fi
-  else
-    rm "${input}.dvconverting"
-    rm -rf ./tmp
```
this is where lockfile and tmpdir cleanup should be

gacopl commented on dvmkv2mp4:
```diff
@@ -111,11 +121,8 @@ function cleanup {
         cp "$i" "${inputbase}$(echo "$i" | sed -n 's/.*\(\.[^.]*\.\(srt\|sup\)\)$/\1/p')"
         echo "${inputbase}$(echo "$i" | sed -n 's/.*\(\.[^.]*\.\(srt\|sup\)\)$/\1/p')"
       done
-      rm "${output}.nfo" "${output}.asm" "${output}.*.srt" "${output}.*.sup" "${input}.dvconverting" 
```
here lockfile removal should stay in case of failed conversion

masupilamie commented:
Thanks for reviewing my commits.

As I see it there are multiple ways how to handle lock files and the temporary directory:

- Have one temporary directory and a lock file that is present during whole script run - this is how I had made it
- Have one temporary directory and a lock file per file conversion - I think this is what you have in mind
- Have a temporary directory per conversion with its own lock file - maybe this is a better solution

The last option is probably the best / safest since there is no risk of temporary files being used for other conversions. This also allows to have a lock per conversion instead of a lock for the whole script run. I've added this logic, it creates a subfolder in the working directory (or --temp-dir/--dest-dir when the option is provided) using file name with '.tmp' as suffix for the subdirectory name. Let me know how you feel about this approach.

I've also removed the --lock-in-temp option since the lock file will now always be in the temporary directory. Lock file will only be removed when debug is not enabled to prevent debug files conflicting with a new run.

gacopl commented:
ok looks ok 
have you tested it properly with without debug etc to make sure files are where they should be and not get deleted :) ?

masupilamie commented:
Yes, tested both with and without debug and using only working directory and with --dest-dir/--temp-dir options active. Did not encounter any unexpected behaviour.
Note: having debug on with multiple files in a directory can generate a lot of used disk space, but I consider this normal debug behaviour.

gacopl commented:
ok i trust you :D thanks for your work

masupilamie commented:
Glad I could contribute to the script. Maybe I have some other improvement ideas but I'll share those via an issue / feature request so it can be discussed before coding it.
